tharsis/resource_managed_identity_access_rules.go

```diff
@@ -0,0 +1,328 @@
+	if req.ProviderData == nil {
+		return
+	}
+	t.client = req.ProviderData.(*tharsis.Client)
+}
+
+func (t *managedIdentityAccessRuleResource) Create(ctx context.Context,
+	req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	// Retrieve values from plan.
+	var plan ManagedIdentityAccessRuleModel
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Build the access rule input.
+	accessRuleInput := ttypes.CreateManagedIdentityAccessRuleInput{
+		ManagedIdentityID:      plan.ManagedIdentityID.ValueString(),
+		RunStage:               ttypes.JobType(plan.RunStage.ValueString()),
+		AllowedUsers:           valueStrings(plan.AllowedUsers),
+		AllowedServiceAccounts: valueStrings(plan.AllowedServiceAccounts),
+		AllowedTeams:           valueStrings(plan.AllowedTeams),
+	}
+
+	// Create the managed identity access rule.
+	created, err := t.client.ManagedIdentity.CreateManagedIdentityAccessRule(ctx,
+		&accessRuleInput)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating managed identity access rule",
+			err.Error(),
+		)
+		return
+	}
+
+	// Map the response body to the schema and update the plan with the computed attribute values.
+	// Because the schema uses the Set type rather than the List type, make sure to set all fields.
+	plan.ID = types.StringValue(created.Metadata.ID)
+	plan.RunStage = types.StringValue(string(created.RunStage))
+	plan.ManagedIdentityID = types.StringValue(created.ManagedIdentityID)
+
+	allowedUsers := []types.String{}
+	for _, user := range created.AllowedUsers {
+		allowedUsers = append(allowedUsers, types.StringValue(user.Username))
+	}
+	plan.AllowedUsers = allowedUsers
+
+	allowedServiceAccounts := []types.String{}
+	for _, serviceAccount := range created.AllowedServiceAccounts {
+		allowedServiceAccounts = append(allowedServiceAccounts, types.StringValue(serviceAccount.ResourcePath))
+	}
+	plan.AllowedServiceAccounts = allowedServiceAccounts
+
+	allowedTeams := []types.String{}
+	for _, team := range created.AllowedTeams {
+		allowedTeams = append(allowedTeams, types.StringValue(team.Name))
+	}
+	plan.AllowedTeams = allowedTeams
+
+	// Set the response state to the fully-populated plan.
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (t *managedIdentityAccessRuleResource) Read(ctx context.Context,
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state ManagedIdentityAccessRuleModel
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the managed identity access rule from Tharsis.
+	found, err := t.client.ManagedIdentity.GetManagedIdentityAccessRule(ctx,
+		&ttypes.GetManagedIdentityAccessRuleInput{
+			ID: state.ID.ValueString(),
+		})
+	if err != nil {
```

Comment:
The `resp.State.RemoveResource` function should be called if the access rule doesn't exist.

Reply:
This applies to all resource types.